java-frontend/src/main/java/org/sonar/java/se/checks/ObjectOutputStreamCheck.java

```diff
@@ -0,0 +1,93 @@
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S2689")
+public class ObjectOutputStreamCheck extends SECheck {
+
+  private static final MethodMatcher FILES_NEW_OUTPUT_STREAM = MethodMatcher.create()
+    .typeDefinition("java.nio.file.Files")
+    .name("newOutputStream")
+    .withAnyParameters();
+
+  public enum FileOutputStreamConstraint implements Constraint {
+    APPEND
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    ProgramState programState = context.getState();
+    if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
+      NewClassTree newClassTree = (NewClassTree) syntaxNode;
+      if (newClassTree.symbolType().is("java.io.FileOutputStream") && newClassTree.arguments().size() == 2) {
+        ProgramState psBeforeInvocation = context.getNode().programState;
+        BooleanConstraint argConstraint = psBeforeInvocation.getConstraint(psBeforeInvocation.peekValue(), BooleanConstraint.class);
+        if (argConstraint == BooleanConstraint.TRUE) {
+          programState = programState.addConstraint(programState.peekValue(), FileOutputStreamConstraint.APPEND);
+        }
+      }
+    } else if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) syntaxNode;
+      if (FILES_NEW_OUTPUT_STREAM.matches(mit)) {
+        ProgramState psBeforeInvocation = context.getNode().programState;
+        int optionsNumber = mit.arguments().size() - 1/*skip Path argument*/;
+        for (int i = 0; i < optionsNumber; i++) {
+          FileOutputStreamConstraint argConstraint = psBeforeInvocation.getConstraint(psBeforeInvocation.peekValue(i), FileOutputStreamConstraint.class);
+          if (argConstraint == FileOutputStreamConstraint.APPEND) {
+            programState = programState.addConstraint(programState.peekValue(), FileOutputStreamConstraint.APPEND);
+          }
+        }
+      }
+    } else if (syntaxNode.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree member = (MemberSelectExpressionTree) syntaxNode;
+      if (member.symbolType().is("java.nio.file.StandardOpenOption") && "APPEND".equals(member.identifier().name())) {
+        programState = programState.addConstraint(programState.peekValue(), FileOutputStreamConstraint.APPEND);
+      }
+    }
+    return programState;
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
+      NewClassTree newClassTree = (NewClassTree) syntaxNode;
+      if (newClassTree.symbolType().is("java.io.ObjectOutputStream")) {
+        ProgramState programState = context.getState();
+        FileOutputStreamConstraint constraint = programState.getConstraint(programState.peekValue(), FileOutputStreamConstraint.class);
+        if (constraint == FileOutputStreamConstraint.APPEND) {
+          context.reportIssue(syntaxNode, this, "This file was opened in append mode.");
```

Comment:
I'm not that much convinced by the message, especially has we are going to have a `FileInputStream` here, not a `File`. No secondaries/flow?